### PR TITLE
Replace filtering API; includes support for filtering numbers

### LINKED ISSUE #385: Implement filtering options for Number types.
## Problem
We need to ensure that records that include columns of all Number types support the following filters via API: 
- between {x} and {y}
- equals {x}
- does not equal {x}
- greater than {x}
- lesser than {x}
- greater than or equal to {x}
- less than or equal to {x}
- is empty
- is not empty

This involves:
- Implementing the filters in the backend
- Updating the `/api/v0/databases/<id>/types/` endpoint to store available filters on this type
  - Filter information should include the number of parameters needing to be passed in (e.g. `between` needs 2 parameters, `is empty` needs 0)

## Additional context
- We're using our fork of `sqlalchemy-filters` to provide filtering. See: https://github.com/centerofci/sqlalchemy-filters.
- #557 provides some context on storing filters.
- This issue is blocked by implementation of Number types: 
  - #377
  - #378 
  - #379 
  - #380 
  - #381 
  - #382 
  - #383 
  - #434

### PR DESCRIPTION
**Pull request by dmos62**
Fixes #385

Related to #1028

This is a continuation of the #1061 PR. Contains 2 major changes:

- replaces current filtering API with the new filtering API (introduced in #1061);
  - new syntax; frontend will use the filters declared in `/api/ui/v0/database/0/filters/`;
  - the `records` query param is now `filter`, not `filters` (singular, not plural);
- it also converts the `duplicate_only` operation from a filter into a query param (see #1028 for some context); to be clear, this had to be done, there wasn't a coherent way to consider it a Mathesar filter.

Also, includes number filters necessary to fulfil requirements in #385. Actually, it's the previous PR, #1061, that included these filters, but now they will be usable.

**Technical details**

Filters are specified by frontend in the function API format:

```python
{"and": [
    {"empty": [
        {"column_reference": ["some_column"]},
    ]},
    {"equal": [
        {"column_reference": ["some_column"]},
        {"literal": ["some_literal_value"]},
    ]},
]}
```

The expected data structure is made up of dicts containing one (and only one) key-value pair each. The key is the filter id, and the value is always a list of parameters. Every parameter is either a nested filter, or a string/number when it's the sole parameter to `literal` or `column_reference`.

There are 5 "special filters" that are not declared with the rest (in `/api/ui/v0/database/0/filters/`) and have to be hardcoded:
- the operators `and`, `or`, `not`; their syntax is equivalent to that of regular filters;
- the `column_reference`, whose parameter is a plain string denoting the column's name;
- the `literal`, whose parameter is some non-collection JSON primitive whose value will be used verbatim (see parameters of the `equal` sub-filter above).

`column_reference` and `literal` "special filters" must always be used when you want to pass a column reference or a literal. The API should throw an error otherwise.

New filter API uses the functions API under-the-hood. Filters are actually `DBFunction`s presented in a simplified way. The syntax for submitting filters to backend is actually just the function API syntax. The filter API could be seen as a thin layer on top of the function. API.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by kgodey** [comment]
> the `column_reference`, whose parameter is a plain string denoting the column's name;

We should be using column ID as the primary way to refer to a column, we're in the process of switching the entire API to do this, see:
- #839 
- #896 
- #898

**Comment by pavish** [comment]
@dmos62 Please let me know when this PR is ready for frontend work. I'll take it up.

**Comment by dmos62** [comment]
@kgodey I'm aware of the move to using column IDs. It's not yet in master though, or is it? Should I refactor this PR to use column ids?

I'll rename `column_reference` to `column_id`.

**Comment by kgodey** [comment]
> It's not yet in master though, or is it?

@silentninja is working on these, hopefully they should be in `master` by the end of the week.

**Comment by dmos62** [comment]
Added support for specifying columns via their Django IDs.

Now the JSON expected by the `records` endpoint's `filter` query param looks like this:

```python
{"and": [
    {"empty": [
        {"column_id": [1]},
    ]},
    {"equal": [
        {"column_id": [3]},
        {"literal": ["some_literal_value"]},
    ]},
]}
```

The difference from the sample posted in the first post is that `column_reference`s are replaced by `column_id`s, which take Django column IDs.

### Details

Made DB layer unaware of Django IDs, by preprocessing the passed DBFunction JSON specification (for example to the `records` endpoint `filter` query param) to rewrite column id references to column name references. Before preprocessing/rewritting, the JSON looks like `dict(empty=[dict(column_id=1)])`, and after it looks like `dict(empty=[dict(column_name="foo")])`. The DB layer only sees the latter (`column_name`) version.

**Comment by kgodey** [comment]
>  {"column_id": [1]},

Why is `column_id` a list? Presumably it's just one `id`.

**Comment by dmos62** [comment]
@kgodey `column_id` is a `DBFunction` and all `DBFunction`s take a sequence of parameters. It's optimized for simple automatic compiling and parsing, not for compilation by-hand. I've considered handling these single-parameter cases differently, but opted not to in order to keep the parsing simple.

**Comment by kgodey** [comment]
@dmos62 To make the API easier to read, I recommend dropping the list from `column_id` and other single parameter options, you can always check if the input is only a single parameter and turn it into a list before passing it on to the `DBFunction`. It doesn't seem like it would be much effort or complicate parsing much.

**Comment by dmos62** [comment]
@kgodey I think this makes the syntax more complicated and thus less robust, but I'll make the change to keep discussions to a minimum.

**Comment by dmos62** [comment]
@kgodey I replied that I'd make the change, but then thought about it some more and came to the conclusion that it's not a good idea.

Having two forms for passing parameters grows the code complexity more than you'd expect, since we're parsing the function JSON spec twice. Once to rewrite IDs to names and again to instantiate a `DBFunction` instance. A change in one parsing routine, has to be kept in sync with the other routine.

I don't think that the readability improvement is significant enough to justify that. I'd like to keep the syntax as simple as possible.

I have considered switching syntax format from `{id: [param1,param2]}` to `{"id": id, "parameters": [param1,param2]}`. That way it's clear that `"parameters"` is always a sequence. I like this for being more orthodox (predictable) JSON, though what's more readable is debatable, since once you understand the current syntax, it's very efficient in terms of information density.

File: db/functions/operations/deserialize.py [hunk 1]
```diff
@@ -45,7 +62,20 @@ def _get_db_function_subclass_by_id(subclass_id):
     for db_function_subclass in known_db_functions:
         if db_function_subclass.id == subclass_id:
             return db_function_subclass
-    raise UnknownDBFunctionId
+    raise UnknownDBFunctionId(
```
Comment (kgodey) [inline db/functions/operations/deserialize.py:65]: `Id` --> `ID`

File: mathesar/functions/operations/convert.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    IDs and can only accept columns referenced by name.
+    """
+    # Using a private method to do the heavy lifting since it uses different parameters.
+    return _rewrite(
+        column_ids_to_names=column_ids_to_names,
+        spec_or_literal=spec
+    )
+
+
+def _rewrite(column_ids_to_names, spec_or_literal):
```
Comment (kgodey) [inline mathesar/functions/operations/convert.py:18]: I only see the JSON spec being parsed here - couldn't you check for single parameters here and convert them to a list during this rewrite?

**Review by kgodey — CHANGES_REQUESTED** [review]
@dmos62 This looks good overall, added some minor comments. I'm fine with the current API spec for now if it's going to result in repeated logic, but I did have a question about why we couldn't put the logic in a single place.

Could you open a separate GitHub issue to reevaluate the filtering API spec with a link to this discussion? It should go in the API v1 milestone (which is post-alpha).

File: mathesar/functions/operations/convert.py (hunk 2 above)
Comment (dmos62) [reply mathesar/functions/operations/convert.py:18]: The other place is `db.functions.operations.deserialize`.

**Comment by dmos62** [comment]
I'll open a ticket to track this filter/functions API spec issue. I'll give it a technical debt label.

**Comment by dmos62** [comment]
@kgodey Renamed `UnknownDBFunctionId` to `UnknownDBFunctionID` (Id -> ID) and removed the obsolete comment. Should be ready for merge.

**Review by kgodey — APPROVED** [review]
Looks good to me. I assume @pavish needs to make frontend changes before we merge?

**Comment by pavish** [comment]
I'll get started on the frontend changes. I'll not be using the `/api/ui/v0/database/${DB}/filters/` endpoint, there's another ticket open to handle that: https://github.com/centerofci/mathesar/issues/1072. In this PR, I'll only fix parts that are breaking.

I have a couple things to confirm:

* @dmos62 I assume [this is the finalized format](https://github.com/centerofci/mathesar/pull/1069#issuecomment-1041558716).
* @kgodey I assume we will have to remove the `has_duplicates` filter option from the UI entirely.
 
Please let me know if these assumptions are correct.

**Comment by kgodey** [comment]
> @kgodey I assume we will have to remove the has_duplicates filter option from the UI entirely.

Yes, we're designing a different way to handle that in #1028. We can create an implementation issue for duplicates once that is done.

**Comment by kgodey** [comment]
@pavish I'm fine with breaking things in `master` temporarily as long as we can make sure that filtering functionality works as expected by the time #1072 is closed. I'll leave the schedule of which things get handled in which PR to you and @seancolsen.

**Comment by kgodey** [comment]
@pavish Does that mean this PR can be merged as-is?

**Comment by dmos62** [comment]
> I'll get started on the frontend changes. I'll not be using the `/api/ui/v0/database/${DB}/filters/` endpoint, there's another ticket open to handle that: #1072. In this PR, I'll only fix parts that are breaking.
> 
> I have a couple things to confirm before I begin:
> 
>     * @dmos62 I assume [this is the finalized format](https://github.com/centerofci/mathesar/pull/1069#issuecomment-1041558716).
> 
>     * @kgodey I assume we will have to remove the `has_duplicates` filter option from the UI entirely.
> 
> 
> Please let me know if these assumptions are correct.

@pavish Yes, your assumptions are correct. I've forgotten to post a summary of that for you, sorry! [This is the final format](https://github.com/centerofci/mathesar/pull/1069#issuecomment-1041558716) and the `duplicate_only` filter is not a filter anymore, but a query param on the `records` endpoint.

This PR shouldn't block any other developers for a while, so I'd say feel free to merge in whichever order you prefer.

@kgodey @pavish @seancolsen Would it make sense to merge into a dedicated branch (that is not master) until it's not broken anymore and then merge that into master, thus not breaking master?

**Comment by kgodey** [comment]
> @kgodey @pavish @seancolsen Would it make sense to merge into a dedicated branch (that is not master) until it's not broken anymore and then merge that into master, thus not breaking master?

I don't think it's worth the fuss at this point since no one is using `master` for anything important. After our alpha release, we should not be breaking `master` anymore.

**Comment by pavish** [comment]
> @pavish Does that mean this PR can be merged as-is?

@kgodey Yes, I think we can merge this PR as-is.

**Pull request merged**